civis/_utils.py

```diff
@@ -59,34 +61,46 @@ def open_session(api_key, max_retries=5, user_agent="civis-python"):
     user_agent = "{}/Python v{} Civis v{} {}".format(
         user_agent, ver_string, civis_version, session_agent)
     session.headers.update({"User-Agent": user_agent.strip()})
-    max_retries = AggressiveRetry(max_retries, backoff_factor=.75,
-                                  status_forcelist=civis.civis.RETRY_CODES)
-    adapter = HTTPAdapter(max_retries=max_retries)
-    session.mount("https://", adapter)
 
     return session
 
 
-class AggressiveRetry(Retry):
-    # Subclass Retry so that it retries more things. In particular,
-    # always retry API requests with a Retry-After header, regardless
-    # of the verb.
-    def is_retry(self, method, status_code, has_retry_after=False):
-        """ Is this method/status code retryable? (Based on whitelists and control
-        variables such as the number of total retries to allow, whether to
-        respect the Retry-After header, whether this header is present, and
-        whether the returned status code is on the list of status codes to
-        be retried upon on the presence of the aforementioned header)
-        """
-        if (self.total and
-                self.respect_retry_after_header and
-                has_retry_after and
-                (status_code in self.RETRY_AFTER_STATUS_CODES)):
-            return True
-
-        else:
-            return super().is_retry(method=method, status_code=status_code,
-                                    has_retry_after=has_retry_after)
+def retry_request(method, prepared_req, session, max_retries=10):
+
+    def _make_request(req, sess):
+        """send the prepared session request"""
+        response = sess.send(req)
+        return response
+
+    def _return_last_value(retry_state):
+        """return the result of the last call attempt
+        and let code pick up the error"""
+        return retry_state.outcome.result()
+
+    if method == 'post':
+        retry_conditions = (
+            retry_if_result(
+                lambda res: res.status_code in civis.civis.POST_RETRY_CODES)
+        )
+    elif method in civis.civis.RETRY_VERBS:
+        retry_conditions = (
+            retry_if_result(
+                lambda res: res.status_code in civis.civis.RETRY_CODES)
+        )
+
+    if retry_conditions:
+        retry_config = Retrying(
+            retry=retry_conditions,
+            wait=wait_for_retry_after_header(
+                fallback=wait_random_exponential(multiplier=2, max=60)),
+            stop=(stop_after_delay(600) | stop_after_attempt(max_retries)),
+            retry_error_callback=_return_last_value,
+        )
+        response = retry_config(_make_request, prepared_req, session)
+        return response
+
+    response = _make_request(prepared_req, session)
+    return response
 
 
 def retry(exceptions, retries=5, delay=0.5, backoff=2):
```

Comment:
Is this going to get torn out or is it still being used?

Reply:
Yeah I'm hesitant to mess with that as its specific to the io portion of the library and is only used there. I added some jitter as an improvement. However the retry logic is a bit more complex than what we're doing with the main client retry. I talked to brennan at one point he he said he kind of whished we could separate these out into two different packages (cli and io).

Reply:
👍